python/src/spark_rapids_ml/umap.py

```diff
@@ -0,0 +1,476 @@
+        must be subsampled to fit within local memory and execute in a reasonable time. Smaller fractions will result in
+        faster training, but may result in sub-optimal embeddings.
+
+    featuresCol: str
+        The name of the column that contains input vectors. featuresCol should be set when input vectors are stored
+        in a single column of a dataframe.
+
+    featuresCols: List[str]
+        The names of the columns that contain input vectors. featuresCols should be set when input vectors are stored
+        in multiple columns of a dataframe.
+
+    labelCol: str (optional)
+        The name of the column that contains labels. If provided, supervised fitting will be performed, where labels
+        will be taken into account when optimizing the embedding.
+
+    Examples
+    --------
+    >>> from spark_rapids_ml.umap import UMAP
+    >>> from cuml.datasets import make_blobs
+    >>> X, _ = make_blobs(1000, 10, centers=42, cluster_std=0.1, dtype=np.float32, random_state=10)
+    >>> df = spark.createDataFrame(X, ["features"])
+    >>> df.show()
+    # TODO: show DF
+    >>> local_model = UMAP(sample_fraction=0.5)
+    >>> local_model.setFeaturesCol("features")
+    >>> distributed_model = local_model.fit(df)
+    >>> embeddings = distributed_model.transform(df)
+    >>> embeddings.show()
+    # TODO: show output DF
+
+    """
+
+    def __init__(self, sample_fraction: float = 1.0, **kwargs: Any) -> None:
+        super().__init__()
+        self.set_params(**kwargs)
+        self.sample_fraction = sample_fraction
+
+    def _create_pyspark_model(self, result: Row) -> _CumlModel:
+        raise NotImplementedError("UMAP does not support model creation from Row")
+
+    def _fit(self, dataset: DataFrame) -> "UMAPModel":
+        if self.sample_fraction < 1.0:
+            data_subset = dataset.sample(
+                withReplacement=False,
+                fraction=self.sample_fraction,
+                seed=self.cuml_params["random_state"],
+            )
+        else:
+            data_subset = dataset
+
+        input_num_workers = self.num_workers
+        # Force to single partition, single worker
+        self._num_workers = 1
+        if data_subset.rdd.getNumPartitions() != 1:
+            data_subset = data_subset.repartition(1)
+
+        pipelined_rdd = self._call_cuml_fit_func(
+            dataset=data_subset,
+            partially_collect=False,
+            paramMaps=None,
+        )
+        rows = pipelined_rdd.collect()
+        # Collect and concatenate row-by-row fit results
+        embeddings = [row["embedding_"] for row in rows]
+        raw_data = [row["raw_data_"] for row in rows]
+        del rows
+
+        model = UMAPModel(
+            embedding_=embeddings,
+            raw_data_=raw_data,
+            n_cols=len(raw_data[0]),
+            dtype=str(np.array(raw_data[0][0]).dtype),
```

Comment:
Would this dtype just be float64 always?

Reply:
I'll fix this. I should just check the python type at this point; getting the numpy version of a python float seems to just default to float64.